### PR TITLE
feat: bootstrap and graceful shutdown

### PR DESCRIPTION
**Pull request by moreirathomas**
## Description

Breaks `main` in many methods for a more granular configuration. Offers:

- configurable graceful shutdown procedure with `shutdown` package.
- `config` struct defining the program variable configuration (à la 12 factors app), config is read from file with default being `.env`. Accepts one single flag: `config` to provide the path to the config file.

## Changes

Refactor `cmd/main`.

Update `golangci-yml`:
- use camelCase check names (for yaml reference intellisense)
- increase min cognitive complexity report from 8 to 15 (recommended is between 10 and 20, so right in the middle)

## Notes

Coses #14

### CODE REVIEW

File: shutdown/signal.go [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package shutdown
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+)
+
+// ListenInterrupt listens for OS interrupt signals and calls f on receive.
+// Calling ListenInterrupt blocks until a signal is received and thus must
```
Comment (c1-ra) [inline shutdown/signal.go:10]: [typo] I think a "be" is missing here ;)

File: cmd/main.go [hunk 2]
```diff
@@ -2,45 +2,122 @@ package main
...
+	defaultConfigPath = ".env"
+	defaultPort       = "9998"
+)
 
 func main() {
-	port := flag.String("port", defaultPort, "Address for the server to listen on.")
-	flag.Parse()
-	addr := ":" + *port
+	ctx, cancel := context.WithCancel(context.Background())
+	go shutdown.ListenInterrupt(cancel)
```
Comment (GregoryAlbouy) [inline cmd/main.go:25]: [question] What's the use case for this? When is the server supposed to receive an interrupt signal? Doesn't it run continuously once deployed? 🤔

File: shutdown/handle.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+// Call calls the registered handlFunc on Handle.
+// Invoking Call is noop if handlFunc is nil.
+func (h *Handle) Call() error {
+	if h.handleFunc == nil {
+		return nil
+	}
+	return h.handleFunc()
+}
```
Comment (GregoryAlbouy) [inline shutdown/handle.go:22]: [change required] What's this? I don't get it: It doesn't provide anything but a complex and error-prone way to only call a function declared by the consumer... Via a method `Call` to be called by the consumer itself!
If it has to both declare and call the function and there's nothing more involved, then what's the point?

```go
func main() {
    f, _ := run()
    if f != nil { // either check for nil or assume it's not
        f.Call() // no idea if myFunc was called at this point
    }
}

func run() (*funccaller.Func, error) { // unusual signature
    // ...
    return funccaller.NewFunc(myFunc), nil
}

func myFunc() {}
```
vs
```go
func main() {
    _ = run()
    myFunc()
}

func run() error {
    // ...
    return nil
}

func myFunc() {}
```

Also `shutdown.Handle` represents how you decided to use it in `main` but it doesn't quite match what it actually does (call any function via a method `Call`): no shutdown-related logics nor handling stuff involved here 🤔

File: cmd/main.go [hunk 4]
```diff
@@ -2,45 +2,122 @@ package main
...
+		addr: ":" + defaultPort,
+	}
+}
+
+func readConfigFile(file string) (config, error) {
+	config := defaultConfig()
+
+	if err := godotenv.Load(file); err != nil {
+		return config, err
+	}
```
Comment (GregoryAlbouy) [inline cmd/main.go:105]: [change required] We can't return here: the server is deployed from the git repository where `.env` is ignored, so we need be able to read injected environment variables without that file.

File: cmd/main.go [hunk 5]
```diff
@@ -2,45 +2,122 @@ package main
...
+	addr       string
+	project    string
+	collection string
+}
+
+func defaultConfig() config {
+	return config{
+		addr: ":" + defaultPort,
+	}
+}
```
Comment (GregoryAlbouy) [inline cmd/main.go:98]: [suggestion] Maybe `defaultConfig` a bit misleading because it returns an invalid config, only the default port is set. I'd suggest to manage the port defaulting inside `readConfigFile` without relying on that function.

File: cmd/main.go [hunk 6]
```diff
@@ -2,45 +2,122 @@ package main
...
+		}
+	}()
+
+	return handle, nil
+}
+
+type config struct {
+	addr       string
+	project    string
+	collection string
```
Comment (GregoryAlbouy) [inline cmd/main.go:91]: [question] Are collection ids / table names really part of a server config? It's not sensitive nor dependent of the environment to me, so it could be hardcoded as well. (Just wondering, not very important here)

File: cmd/main.go [hunk 7]
```diff
@@ -2,45 +2,122 @@ package main
...
+}
+
+// run executes the main program. It starts a server.Server inside
+// a goroutine and returns a configured shutdown.Handle for gracefully
+// shutting down the program.
+//
+// If the bootstrap of the server fails, run returns a non-nil error.
+// Otherwise the caller must use the shutdown.Handle to stop the server.
+func run(ctx context.Context, cancel context.CancelFunc) (*shutdown.Handle, error) {
+	configPath := flag.String("config", defaultConfigPath, "Path to the configuration file (for example --config .env)")
```
Comment (GregoryAlbouy) [inline cmd/main.go:50]: [nitpick] It's more an environment file than a server config to me
```suggestion
	envFile := flag.String("envFile", defaultEnvFile, "Path to environment file (for example -envFile .env)")
```

File: cmd/main.go [hunk 8]
```diff
@@ -2,45 +2,122 @@ package main
...
-	collectionID := os.Getenv("FIRESTORE_COLLECTION_ID")
-	if projectID == "" {
-		log.Fatalf("FIRESTORE_COLLECTION_ID variable is not defined")
+	config, err := readConfigFile(*configPath)
+	if err != nil {
+		return &shutdown.Handle{}, err
 	}
 
-	rs, err := firestore.NewReportService(context.Background(), projectID, collectionID)
+	rs, err := firestore.NewReportService(context.Background(), config.project, config.collection)
```
Comment (GregoryAlbouy) [inline cmd/main.go:59]: [suggestion]
```suggestion
	rs, err := firestore.NewReportService(ctx, config.project, config.collection)
```

**Review by GregoryAlbouy — CHANGES_REQUESTED** [review]
I'm sorry I had a hard time understanding this PR, starting with its purpose but also regarding some implementation choices.

I find the articulation `main`-`run` more complex than it should be, and I think this is partly due to that weird thing with `shutdown.Handle` 😕

**Comment by moreirathomas** [comment]
The main thing is: 

In order to be able to call `Server.Shutdown(ctx)` and `RequestService.Close()`, we need to have `Server.ListenAndServe` to not be blocking.

`Server.Shutdown(ctx)` must be called while `Server.ListenAndServe` runs in a goroutine.

Only after `Server.ListenAndServe` internally gets the interrupt and return `http.ErrServerClosed` and we can safely exit the program.

My solution was to:
- put `Server.ListenAndServe` inside a goroutine
- return the callback to gracefully stop the server when any interrupt signal is received (manual ctrl+c or some redeployment trigger from our hosting service)

In other words: `run` launches the server in a thread and returns a way to gracefully stop it from the parent. Any error returned is a fatal error during launch.

Although I agree that `shutdown.Handle` may be too much.

--- 
Also, you raised good points regarding the env vars injections on real world deployment too. We might need to think about that.
